R/data.table.R

```diff
@@ -1121,6 +1121,11 @@ replace_dot_alias = function(e) {
           if (is.name(lhs)) {
             lhs = as.character(lhs)
           } else {
+            #6033 revdep. Slowly deprecate in 1.17.0. Caller has given us `dt[, substitute(names(.SD))]` which means
+            # jsub is actually substitute(names(.SD)) instead of just names(.SD)
+            if (lhs %iscall% 'substitute')
+              lhs = eval(lhs, parent.frame(), parent.frame())
```

Comment:
do we need to double-`eval()`? Or can we do `if (lhs %iscall% 'substitute') lhs = eval() else lhs = eval()`?

Reply:
That's my understanding which is why I never understood how it used to work.
``` r
my_col = "a"
jsub = substitute(substitute(my_col))
eval(jsub, parent.frame(), parent.frame())
#  my_col
eval(eval(jsub))
# [1] "a"
```
At the same time, there's this head scratcher:
``` r
eval(jsub, list(my_col = "b"), parent.frame())
# [1] "b"
```

Reply:
Head scratcher indeed... OK I'm more of the opinion we shouldn't allow `j=substitute(...)`, indeed it's hard to reason about what `jsub = substitute(subsitute(...))` might do. We can start a deprecation process & see if there's a good use case, my guess is, the end user would be better served by some other approach though.